### PR TITLE
[move-vm] refactor the module loading logic

### PR DESCRIPTION
**Pull request by mengxu-fb**
This commit is the first step towards refactoring the module loading
logic in the Move VM loader.

We model the dependency relations of the modules in the storage as a
directed graph G:
- each module is a node
- There are two types of edges G:
  - Dep edge: A -d-> B, means that module A depends on B (by calling a
    function in B or using a type defined in B)
  - Friend edge: A -f-> B, means that module A declares that B is a
    friend of A

We further define some notion about G:
- A dep closure of module A, denoted as D(A), is the set of modules in
  G that are reachable from A by following **only** the dep edge -d->
- A complete closure of module A, denoted as C(A), is the set of modules
  in G that are reachable from A by following both the dep edge and the
  friend edge.
- Additionally, we define that if a module F is a immediate friend of
  any module in D(A) and F does not belong to D(A), then, F is in the
  **friendship frontier** of D(A). The friendship frontier is a set of
  modules like F and is denoted as FF(D(A)).

The code cache will maintain the following invariant:
- At the end of a module loading session, if a module A is found in the
  code cache, then
  1) all modules in C(A) are in the code cache
  2) all modules in C(A) pass the bytecode verifier
  3) all modules in C(A) pass the linking check
  4) all modules in C(A) do not form a cycle from the view point of the
     cyclic dependency checker (NOTE: different from the notion here,
     the cyclic dependency checker actually reverses the direction of
     the friend edge when checking for cycles.)
- If any of the property does not hold, a VMError will be returned and
  the code cache will be in a state that 2) and 3) continue to hold,
  there is no guarantee that 1) and 4) will hold. (Even in 2 and 3, the
  C(A) is a subset of the actual closure).

The core logic can be simplified as following:
```
procedure load_module_closure(A):
  load all modules in D(A) into code cache
  calculate FF(D(A))
  for each module F in FF(D(A)):
    load_module_closure(F)

end of procedure if successful: all modules in C(A) are in code cache
```

This logic is encapsulated in the following three functions:

- `load_and_verify_module`
  - Load, deserialize, and check the module with the bytecode verifier.
    All checks are performed with regard to the module itself.
  - If successful, return the `CompiledModule`

- `load_and_verify_module_and_dependencies`
  (together with a co-recursive function `load_and_verify_dependencies`)
  - This function recursively finds and loads all modules in D(A) into
    the code cache.

- `load_and_verify_module_and_dependencies_and_friends`
  (together with a co-recursive function `load_and_verify_friends`)
  - This function recursively calculates FF(D(A)) and loads all modules
    in the new friendship frontier into the code cache. Until the
    friendship frontier has no modules. When this happens, all modules
    in C(A) are in the code cache.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Tests for bad/inconsistent storage states are scattered across several testsuites, including
- move-vm-integration-test
- e2e-testsuite
- move-cli tests

None of the existing tests should fail. Furthermore, pay attention to the E2E tests in CI, they should complete within 35 minutes (normally between 29-31 minutes).

### CODE REVIEW

**Review by modocache — COMMENTED** [review]
I'm not the kind of VM expert that ought to be approving a change like this, but based on my understanding I think this is great! I learned a lot from your pull request description, enough so that I would even recommend writing it as a code comment at the top of the file, so that anyone who reads through the loader source code can understand the series of functions that it is recursively calling and why.

**Comment by vgao1996** [comment]
Question about the algorithm and the definition of D(A), C(A) and FF(D(A)):
If I'm understanding correctly, right now we're being as eager as we can, pulling all modules into the code cache, then we do we need to differentiate the dep edges and friend edges? Can't we just grab transitive closure reachable through either type of the edges? Is it because we need to allow cycles with friend edges?

**Comment by mengxu-fb** [comment]
> Can't we just grab transitive closure reachable through either type of the edges?

This is actually what we are doing at the end of the loading, i.e., all the modules in the transitive closure are in the code cache. The reason we follow the algorithm 

But that said, thanks for the question! I forgot to explain why we are using the algorithm of D(A), C(A), etc, instead of using this simple approach. It is mainly for performance reasons, to be specific, to preserve the order in which modules are loaded into the code cache.

If we grab everything in the transitive closure into memory first, we still need to run a procedure to sort the modules in dependency order and then insert them to code cache. In the algorithm described above, the exploration process preserves the dependency order (i.e., the exploration process is the algorithm to sort modules in order), as a result, whenever a module finishing linking check, that module can be inserted into the code cache already.

**Comment by vgao1996** [comment]
> If we grab everything in the transitive closure into memory first, we still need to run a procedure to sort the modules in dependency order and then insert them to code cache.
I see. And I assume this is because items in a module need to be globalized before those in the ones that are depending on it.

Nice consideration!

File: language/move-vm/runtime/src/loader.rs [hunk 1]
```diff
@@ -492,7 +492,7 @@ impl Loader {
     fn deserialize_and_verify_script(
         &self,
         script: &[u8],
-        data_store: &mut impl DataStore,
+        data_store: &impl DataStore,
```
[the remark this answers is not in the document]
Comment (mengxu-fb) [reply language/move-vm/runtime/src/loader.rs:495]: Ha, I just realized that this has been address in #8839. Thank you @runtian-zhou!

File: language/move-vm/runtime/src/loader.rs [hunk 2]
```diff
@@ -902,183 +800,245 @@ impl Loader {
...
-    fn load_module_verify_not_missing(
+    // Everything in `load_and_verify_module` and also recursively load and verify all the
+    // dependencies of the target module.
+    fn load_and_verify_module_and_dependencies(
         &self,
         id: &ModuleId,
-        linking_in_progress: &mut BTreeMap<ModuleId, CompiledModule>,
+        bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
         data_store: &impl DataStore,
+        visited: &mut BTreeMap<ModuleId, Vec<ModuleId>>,
```
Comment (mengxu-fb) [inline language/move-vm/runtime/src/loader.rs:888]: Split into:

```
visited: &mut BTreeSet<ModuleId>,
friends_discovered: &mut BTreeSet<ModuleId>,
```

**Comment by vgao1996** [comment]
> It is mainly for performance reasons, to be specific, to preserve the order in which modules are loaded into the code cache.

Actually, on a second thought -- while I still fully agree that the traversal algorithm should preserve the order, I'm wondering if we can decouple module discovery (traversal) and verification.

Let's say we have the following graph:
```
A
|
v
B => C -> E
     |  
     v
     D

# -> depending on
# => friend
```
The current algorithm will 
1. Try to load and verify B, and checks that friend C exists
2. load and verify A
3. load and verify D
4. load and verify E
5. load and verify C

Could we instead have a traversal procedure that simply gathers a (generalized) topological sort of the modules:
```
[B, A, D, E, C]
```
And then try to verify and globalize each of the modules in sequence?

I think it was brought up in the meeting that it was kinda weird that we were interleaving the loading and verification process, and I wonder if the alternative algorithm I proposed would cut down some of its complexities.

**Review by vgao1996 — APPROVED** [review]
But yeah, as we just discussed offline, what I proposed could be considered as a possible next step, and implemented as a followup PR if it ends up making sense. The current implementation is pretty solid and is a lot better than what we previously have so let's get it in first!

**Pull request merged**